src/object.rs

```diff
@@ -386,6 +386,30 @@ glib_wrapper! {
     }
 }
 
+#[macro_export]
+macro_rules! glib_weak_impl {
+    ($name:ident, $crate::wrapper::Void) => {
+        #[doc(hidden)]
+        impl $crate::clone::Downgrade for $name {
+            type Target = $crate::object::WeakRef<Self>;
+
+            fn downgrade(&self) -> Self::Target {
+                <Self as $crate::object::ObjectExt>::downgrade(&self)
+            }
+        }
+
+        #[doc(hidden)]
+        impl $crate::clone::Upgrade for $crate::object::WeakRef<$name> {
+            type Target = $name;
+
+            fn upgrade(&self) -> Option<Self::Target> {
+                self.upgrade()
+            }
+        }
+    };
+    ($name:ident, $ty:ty) => ();
```

Comment:
Why this case? Why do you only generate this for types that have an FFI class struct? That seems wrong, you need to generate it for all types and there's not really any relationship between having an FFI class struct and being able to have weak references.

Reply:
I thought I read somewhere that only final types don't have a FFI class.

Reply:
That's correct but why do you only want to do this for final types? That would mean that you couldn't use the macro for `gio::Cancellable` for example.